### PR TITLE
Add Squad Eval

### PR DESCRIPTION
**Pull request by dlrudwo1269**
As title. This speeds down the eval quite a bit, but should not be significant as the validation set is small (~5 min to go through validation set on CPU)

Sample output: [Val] Epoch    0 | val_loss: 7604.2744 | exact_math: 0.0253 | F1: 3.8502

### CODE REVIEW

File: train_squad.py [hunk 1]
```diff
@@ -91,6 +101,68 @@ def forward_batch(batch):
...
+                    max_score = score
+                    best_range = (start_idx, end_idx)
+        if best_range == (0,0):
+            prediction = {"id": ids[b], "prediction_text": "", "no_answer_probability": 1.0}
+        else:
+            prediction_text = tokenizer.decode(input_ids[b][best_range[0]:best_range[1]+1])
+            prediction = {"id": ids[b], "prediction_text": prediction_text, "no_answer_probability": 0.}
+        predictions.append(prediction)
+
+    return predictions
```
Comment (lbaierreinio) [inline train_squad.py:163]: A couple of questions if you have time. No need to get back to me because I'm mainly just trying to add something to comment before I approve this.
1) We're either setting no_answer_probability to 0 or 1 but I'm guessing it's defined as a probability because it can assume a value between 0 and 1 in certain cases. Is this something we need to consider when computing the squad_matrics?
2) I guess this solution rewards the model equally whether the best prediction range was the 1st or the 20th top prediction. Not sure if we would want to add something to track the fact that the correct prediction wasn't necessarily the model's top one

**Review by lbaierreinio — APPROVED** [review]
Looks good to me! Not completely familiar with the nuances of the SQUAD dataset so hard for me to give too much meaningful feedback but lmk if there's anything specific you want me to go over.

File: train_squad.py (hunk 1 above)
Comment (dlrudwo1269) [reply train_squad.py:163]: > We're either setting no_answer_probability to 0 or 1 but I'm guessing it's defined as a probability because it can assume a value between 0 and 1 in certain cases. Is this something we need to consider when computing the squad_matrics?
By default, the metric evaluation will only take examples explicitly set to `no_answer_probability=1.0` to have no answer. We can adjust that threshold by setting the `no_answer_threshold` key in the ground truth reference, but in our case, it is not necessary as we are directly calculating whether or not the question is answerable Reference: https://github.com/huggingface/evaluate/blob/main/metrics/squad_v2/squad_v2.py

> I guess this solution assesses the model equally whether the best prediction range was the 1st or the 20th top prediction. Not sure if we would want to add something to acknoledge the fact that the best prediction may not have necessarily been the model's top one
It is a bit difficult to determine the "top pair" because there are many invalid combinations. And this is not used for computing the loss / training, but just to get a valid prediction, so I think it is fine
